source/CScriptEngine.cpp

```diff
@@ -347,6 +348,289 @@ namespace CLEO
+            }
+            case DT_VAR_ARRAY:
+                ReadArrayInformation(script, &arrVarOffset, &arrElemIdx);
+                *(SCRIPT_VAR*)(&scmBlock[arrVarOffset + 4 * arrElemIdx]) = opcodeParams[i];
+                break;
+            case DT_LVAR_ARRAY:
+                ReadArrayInformation(script, &arrVarOffset, &arrElemIdx);
+                *GetPointerToLocalVariable(script, arrVarOffset + arrElemIdx) = opcodeParams[i];
+                break;
+            case DT_PUSH:
+                GetInstance().StackPush(opcodeParams[i].dwParam);
+                break;
+            }
+        }
+    }
+
+    // wrapper around CRunningScript::StoreParams to preserve the value of ecx register
+    static void __declspec(naked) HOOK_CRunningScript__StoreParams()
+    {
+        _asm
+        {
+            push ecx                    // save ecx
+            push dword ptr[esp + 8]       // count
+            push ecx					// script
+            call CRunningScript::StoreParams
+            pop ecx                     // restore ecx     
+            retn 4
+        }
+    }
+
+    SCRIPT_VAR* CRunningScript::GetPointerToLocalArrayElement(CRunningScript* script, int off, WORD idx, BYTE mul)
+    {
+        int final_index = off + mul * idx;
+
+        if (script->IsMission())
+            return &missionLocals[final_index];
+        return script->GetVarPtr(final_index);
+    }
+
+    DWORD CRunningScript::CollectNextParameterWithoutIncreasingPC(CRunningScript* script) {
+        WORD arrVarOffset;
+        int arrElemIdx;
+
+        SCRIPT_VAR result;
+        result.dwParam = -1;
+
+        BYTE* ip = script->GetBytePointer();
+
+        int dt = script->ReadDataByte();
+        switch (dt) {
+        case DT_DWORD:
+            result.dwParam = script->ReadDataInt();
+            break;
+        case DT_VAR:
+        {
+            WORD index = script->ReadDataVarIndex();
+            result = *reinterpret_cast<SCRIPT_VAR*>(&scmBlock[index]);
+            break;
+        }
+        case DT_LVAR:
+        {
+            WORD index = script->ReadDataVarIndex();
+            result = *GetPointerToLocalVariable(script, index);
+            break;
+        }
+        case DT_BYTE:
+            result.dwParam = script->ReadDataByte();
+            break;
+        case DT_WORD:
+            result.dwParam = script->ReadDataWord();
+            break;
+        case DT_FLOAT:
+            result.fParam = script->ReadDataFloat();
+            break;
+        case DT_VAR_ARRAY:
+            ReadArrayInformation(script, &arrVarOffset, &arrElemIdx);
+            result = *reinterpret_cast<SCRIPT_VAR*>(&scmBlock[arrVarOffset + (4 * arrElemIdx)]);
+            break;
+        case DT_LVAR_ARRAY:
+            ReadArrayInformation(script, &arrVarOffset, &arrElemIdx);
+            result = *GetPointerToLocalVariable(script, arrVarOffset + arrElemIdx);
+            break;
+        case DT_POP:
+            break;
+        }
+
+        script->SetIp(ip);

```

Comment:
When rewinding please make sure `GetInstance().OpcodeSystem.handledParamCount` is reverted to correct state too. It is used for debug messages. Maybe there is way to handle it some smarter way.